src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/DSAOpenSsl.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Cryptography
+{
+    public sealed partial class DSAOpenSsl : DSA
+    {
+        public DSAOpenSsl(DSAParameters parameters)
+        {
+            ImportParameters(parameters);
+        }
+
+        /// <summary>
+        /// Create an DSAOpenSsl from an <see cref="SafeEvpPKeyHandle"/> whose value is an existing
+        /// OpenSSL <c>EVP_PKEY*</c> wrapping an <c>DSA*</c>
+        /// </summary>
+        /// <param name="pkeyHandle">A SafeHandle for an OpenSSL <c>EVP_PKEY*</c></param>
+        /// <exception cref="ArgumentNullException"><paramref name="pkeyHandle"/> is <c>null</c></exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="pkeyHandle"/> <see cref="Runtime.InteropServices.SafeHandle.IsInvalid" />
+        /// </exception>
+        /// <exception cref="CryptographicException"><paramref name="pkeyHandle"/> is not a valid enveloped <c>DSA*</c></exception>
+        public DSAOpenSsl(SafeEvpPKeyHandle pkeyHandle)
+        {
+            if (pkeyHandle == null)
+                throw new ArgumentNullException(nameof(pkeyHandle));
+            if (pkeyHandle.IsInvalid)
+                throw new ArgumentException(SR.Cryptography_OpenInvalidHandle, nameof(pkeyHandle));
+
+            // If dsa is valid it has already been up-ref'd, so we can just use this handle as-is.
+            SafeDsaHandle key = Interop.Crypto.EvpPkeyGetDsa(pkeyHandle);
+            if (key.IsInvalid)
```

Comment:
We should still Dispose of the SafeHandle even if it's invalid.  Otherwise it'll still be finalized.